gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -0,0 +1,897 @@
+                if (m_poFp != nullptr) {
+                    VSIFCloseL(m_poFp);
+                    m_poFp = nullptr;
+                }
+                return nullptr;
+            }
+        }
+
+        if (m_processedSpatialIndex && !m_ignoreSpatialFilter)
+            m_offset = m_offsetInit + m_featureOffsets[m_foundFeatureIndices[m_featuresPos]];
+        else if (m_featuresPos > 0)
+            m_offset += m_featureSize + sizeof(uoffset_t);
+
+        OGRFeature *poFeature = new OGRFeature(m_poFeatureDefn);
+
+        VSIFSeekL(m_poFp, m_offset, SEEK_SET);
+        VSIFReadL(&m_featureSize, sizeof(uoffset_t), 1, m_poFp);
+        CPL_LSBPTR32(&m_featureSize);
+        if (m_featureBufSize == 0) {
+            m_featureBufSize = std::max(1024U * 32U, m_featureSize);
+            CPLDebug("FlatGeobuf", "GetNextFeature: m_featureBufSize: %d", m_featureBufSize);
+            m_featureBuf = static_cast<GByte *>(VSIMalloc(m_featureBufSize));
+        } else if (m_featureBufSize < m_featureSize) {
+            m_featureBufSize = std::max(m_featureBufSize * 2, m_featureSize);
+            CPLDebug("FlatGeobuf", "GetNextFeature: m_featureBufSize: %d", m_featureBufSize);
+            m_featureBuf = static_cast<GByte *>(VSIRealloc(m_featureBuf, m_featureBufSize));
+        }
+        VSIFReadL(m_featureBuf, 1, m_featureSize, m_poFp);
+
+    #ifdef DEBUG
+        const uint8_t * vBuf = const_cast<const uint8_t *>(reinterpret_cast<uint8_t *>(m_featureBuf));
+        Verifier v(vBuf, m_featureSize);
+        auto ok = VerifyFeatureBuffer(v);
+        if (!ok) {
+            CPLDebug("FlatGeobuf", "VerifyFeatureBuffer says not ok");
+            CPLDebug("FlatGeobuf", "m_offset: %zu", m_offset);
+            CPLDebug("FlatGeobuf", "m_featuresPos: %zu", m_featuresPos);
+            CPLDebug("FlatGeobuf", "featureSize: %d", m_featureSize);
+        }
+    #endif
+
+        auto feature = GetRoot<Feature>(m_featureBuf);
+        auto fid = feature->fid();
+        poFeature->SetFID(fid);
+        //CPLDebug("FlatGeobuf", "GetNextFeature: fid: %zu", fid);
+        auto ogrGeometry = readGeometry(feature);
+        if (ogrGeometry == nullptr) {
+            CPLError(CE_Failure, CPLE_AppDefined, "Failed to read geometry");
+            return nullptr;
+        }
+    #ifdef DEBUG
+        //char *wkt;
+        //ogrGeometry->exportToWkt(&wkt);
+        //CPLDebug("FlatGeobuf", "readGeometry as wkt: %s", wkt);
+    #endif
+        if (m_poSRS != nullptr)
+            ogrGeometry->assignSpatialReference(m_poSRS);
+        if (!m_poFeatureDefn->IsGeometryIgnored())
+            poFeature->SetGeometry(ogrGeometry);
+
+        auto properties = feature->properties();
+
+        if (properties != nullptr) {
+            auto data = properties->data();
+            auto size = properties->size();
+            //CPLDebug("FlatGeobuf", "properties->size: %d", size);
+            uoffset_t offset = 0;
+            while (offset < (size-1)) {
```

Comment:
shouldn't the check take into account the number of bytes that will be actually read ?

Reply:
I'm not sure what you mean here. The properties buffer is a "dumb" binary buffer containg the individual property values prefixed by the column index. The binary length of a value depend on column type for that index. So, the number of bytes to read/parse for each property is not known up front.

Reply:
what I mean is that at line 519 you'll dereference 4 more bytes, so the check `offset < (size-1)` by itself is not enough to guarantee you won't run outside of the buffer. There must be some external guarantee that it won't happen ? I'm considering here the case of corrupted input data

Reply:
I see, I've added checks at each progression of the buffer parse and a check for column index range.